### ISSUE TITLE
The loss_label funciton has a bug!

### ISSUE THREAD

**Issue by AlanZTY** [facebookresearch/detr#164]
target_classes = torch.full(src_logits.shape[:2], self.num_classes,
                                    dtype=torch.int64, device=src_logits.device)
target_classes[idx] = target_classes_o

I believe it should be:
target_classes[idx] = target_classes_o - 1

The reason that this still works is because that same mistake is made when converting logits to class. 

The problem is now the 0th logit is never used. And the 91st class is never represented in the logits. As 91st is supposed to be on 90th logit, but due to the bug it is now on 91st logit. However, 91st is used to represent non-object class.

**Comment by alcinos** [comment]
Hi @AlanZTY 
Thank you for your interest in DETR.

This is not really a bug, we chose this implementation for simplicity reasons. Note that class 0 is traditionally reserved for "background" in many detector implementations, hence we leave it unused to avoid confusion. We use class 91 instead for our "no-object" class. 
Note that even if we shift the labels by 1 as you suggest, this still only partially addresses the problem: the class ids in COCO are sparse (they are in [1, 90] but in reality there are only 80 classes), so you'll end-up with unused logits anyways (which is fine, it's only incurring a minimal waste of parameters). A "real" solution would involve remapping the classes into [0, 79] (as done in Detectron2), but we decided not to do it for code readability.

I believe I have answered your question, and as such I'm closing this. Feel free to reach out if you have further concerns.

**Comment by AlanZTY** [comment]
Thanks! I now understand why you did that. That was a very clear explanation from you. 
I was fine tuning on pedestrians so I only try to use binary cross entropy and I used the wrong logit (should be using 1st but used 0th instead). 
Anyways, many thanks.

**Comment by AlanZTY** [comment]
Hi, Can I ask a question?
The input image has different sizes, to make a batch the same size, you have used a nested tensor funciton. Now the smaller images are bigger padded with zeros. However, the target bboxes have not been changed. 

Could you please explain that why the network still predicts the cx, cy related to original image size?
Is this due to the attention mechanism?

**Comment by fmassa** [comment]
@AlanZTY in https://github.com/facebookresearch/detr/issues/70#issuecomment-641221598 we add some more information about how we handle this.

Here are a few pointers:
- the positional encodings take the original image size into account, and not the padded regions
- we pass a mask to the transformer to the regions that are padded, so that it doesn't take it into account

**Comment by AlanZTY** [comment]
Thanks for your answer. I would like to discuss another question. Do you think it is possible to only use the encoder for detection? Like classification and bbox estimation right after encoder?

**Comment by alcinos** [comment]
Our approach is centered around object queries, which implies the presence of a decoder.
Other approaches might be possible, but likely will differ from DETR in substantial ways.

**Comment by AlanZTY** [comment]
Thanks for you answer! I was training the detr model. For a 30g gv100, I can use a batch size of 32. But for a 11g ti gpu, I can only use a batchsize of 3. So even if I use 3 gpus, I can only use batchsize of 9, which is significantly lower than a single gpu. Could you please help me explain the problem?

**Comment by AlanZTY** [comment]
Sorry I wasn't making sense. Don't bother.
